### PR TITLE
Add macos-14 as a runner (Apple M1)

### PR DESCRIPTION
**Pull request by alalazo**
Requires #43363

This should allow us to run test Apple M1 in CI.

### CODE REVIEW

**Comment by alalazo** [comment]
@spackbot fix style

File: lib/spack/spack/test/relocate.py [hunk 1]
```diff
@@ -282,6 +284,7 @@ def test_relocate_text_bin_raise_if_new_prefix_is_longer(tmpdir):
 
 
 @pytest.mark.requires_executables("install_name_tool", "file", "cc")
+@pytest.mark.skipif(str(archspec.cpu.host().family) != "x86_64", reason="failing on Apple M1/M2")
```
Comment (alalazo) [inline lib/spack/spack/test/relocate.py:287]: I would merge this PR as is, and fix this test in a following PR. The issue here is that `sonoma` takes care of deduplicating rpaths during library creation, so the test is "wrong" in some implementation assumptions.

**Review by adamjstewart — APPROVED** [review]

**Review by becker33 — APPROVED** [review]

**Comment by becker33** [comment]
I'm very surprised there aren't more failures here, I'm seeing 25 failures when I run on my mac. Trying with your branch to see if that's consistent for me, but the changes all look good.

**Comment by alalazo** [comment]
@becker33 I merged this one today too https://github.com/spack/spack/pull/43363 Are you up to date with `develop`?

**Comment by becker33** [comment]
I'm just now testing with the newest develop -- seeing at least one error that's not showing here, but I suspect that may be due to my development environment. I'll wait until the full suite finishes before giving more detail.

**Comment by becker33** [comment]
I'm getting 6 failures on this branch with my machine, and they're all with the tag tests. I think they're due to my environment, so let's merge this as-is and I'll follow up on those and create a PR if they're not caused on my end.

**Comment by becker33** [comment]
I can confirm the failures were specific to my environment. After `spack -m clean -m` they disappeared, my guess is that we had a past bug in which values were not deduped in the tag index, and my index hadn't been cleaned up since before that was fixed.

**Pull request merged**